source/pele/xy_model_online.h

```diff
@@ -0,0 +1,232 @@
+    {
+        if (x.size() != y.size()) {
+            throw std::runtime_error("graph: illegal input: x.size() != y.size()");
+        }
+        for (size_t i = 0; i < x.size(); ++i) {
+            insert_edge(x[i], y[i], directed());
+        }
+        if (nvertices_input > maxv()) {
+            throw std::runtime_error("graph: illegal input: too many vertices");
+        }
+        nvertices_ = nvertices_input;
+    }
+    void insert_edge(const unsigned int x, const unsigned int y, bool no_mirror)
+    {
+        edgenode* p = new edgenode(y, edges.at(x));
+        edges.at(x) = p;
+        ++degree.at(x);
+        if (!no_mirror) {
+            insert_edge(y, x, true);
+        }
+        else {
+            ++nedges_;
+        }
+    }
+    void resize(const unsigned int maxv_)
+    {
+        edges.assign(maxv_, NULL);
+        degree.assign(maxv_, 0);
+        nvertices_ = 0;
+        nedges_ = 0;
+    }
+    void print(std::ostream& stm)
+    {
+        stm << "number of vertices in graph: " << nvertices() << "\n";
+        stm << "number of edges in graph: " << nedges() << "\n";
+        stm << "adjacency list:\n";
+        for (unsigned int i = 0; i < nvertices_; ++i) {
+            stm << i << " ==>> ";
+            edgenode* p = edges.at(i);
+            unsigned int tmp = 0;
+            while (p != NULL) {
+                if (tmp++) {
+                    stm << " --> ";
+                }
+                stm << p->y;
+                p = p->next;
+            }
+            stm << "\n";
+        }
+    }
+    unsigned int nvertices() const
+    {
+        return nvertices_;
+    }
+    unsigned int nedges() const
+    {
+        return nedges_;
+    }
+    bool directed() const
+    {
+        return directed_;
+    }
+    unsigned int maxv() const
+    {
+        return edges.size();
+    }
+    edgenode* get_edges(const unsigned int v) const
+    {
+        return edges.at(v);
+    }
+};
+
+/**
+ * XY model implementation where the energy is computed by summing each
+ * linked list in the adjacency list vector in tun. 
+ */
+class XYModelOnline : public BasePotentialOnline {
+    graph m_topology;
+public:
+    XYModelOnline(const size_t nr_spins, pele::Array<size_t> head, pele::Array<size_t> tail)
+        : BasePotentialOnline(nr_spins),
+          m_topology(nr_spins, head, tail)
+    {}
+    double get_energy(Array<double> x, const size_t index)
+    {
+        /**
+         * Return ith term of the potential energy, i.e., the sum of 
+         * -cos(x[i] - x[j]) for all j which are linked to i
+         */
+        if (x.size() != m_topology.nvertices()) {
+            throw std::runtime_error("XYModelOnline: x.size() != nr vertices in graph");
+        }
+        double energy = 0;
+        edgenode* en = m_topology.get_edges(index);
+        while (en) {
+            energy += -std::cos(x[index] - x[en->y]);
+            en = en->next;
+        }
```

Comment:
since iterating through the out edges of a node is the primary use here, it might be faster to store the out edges in a vector (see below).  

"""
class graph {
     int num_nodes;
    std::vector<std::vector<uint> > connected_nodes;  
}
"""

This would be slower to build the graph, but faster to iterate over because the connected nodes are all co-located in memory.  Just an option